### PR TITLE
Fix #621 Add param mp to stumpi

### LINKED ISSUE #621: Allow user to use pre-calculated `mp` in `stumpy.stumpi.py`
It seems that  `stumpy.stumpi.py` needs to calculate `matrix profile (mp)` in the beginning of the algorithm.

https://github.com/TDAmeritrade/stumpy/blob/576de379994896063be71d264e48a08ae17476cf/stumpy/stumpi.py#L118

And, then it will be updated quickly for each new individual, scaler value `t` (because, I think it just needs to calculate a new distance profile ~for~ considering `t` and update matrix profile.

But, what if the initial `T` is big? In that case, user may want to use `stumped` or `gpu_stump` to get `mp`, and then pass it to `stumpy.stumpi.py`. Or, maybe they store it somewhere and decide to use it later.

So, I was wondering if it would be reasonable to do this:
```
def __init__(..., mp=None):
        """
        mp: numpy.ndarray, default None
        the matrix profile of `T`. If user haven't calculated it yet, set it to None. Otherwise, it is user's responsibility to make sure the input 
        `mp` is  correct.
        """
        if mp is None:
             mp = stump(self._T, self._m) # user can now calculate the `mp` using `gpu_stump`
```

### PR DESCRIPTION
**Pull request by NimaSarajpoor**

### CODE REVIEW

File: stumpy/aampi.py
Comment (seanlaw) [inline stumpy/aampi.py:39]: You've jumped right into the description but you should really start off with "A pre-computed matrix profile (and corresponding matrix profile indices). This is a 2D array..."

File: tests/naive.py
Comment (seanlaw) [inline tests/naive.py:926]: use `self._T, self._m`

File: stumpy/aampi.py
Comment (seanlaw) [inline stumpy/aampi.py:129]: See comment below

File: stumpy/aampi.py
Comment (NimaSarajpoor) [reply stumpy/aampi.py:39]: Great point! Thanks for the input. I will revise the docstrings.

File: stumpy/stumpi.py
Comment (NimaSarajpoor) [reply stumpy/stumpi.py:163]: I was trying to help user understand the components of shape. But...this is not a place for tutorial :) I will revise it.

Comment (seanlaw) [reply stumpy/stumpi.py:163]: It's not a bad idea but the immediate message should really be that this `mp` doesn't appear to be for this `T`

Comment (NimaSarajpoor) [reply stumpy/stumpi.py:163]: @seanlaw 
>  but the immediate message should really be that this mp doesn't appear to be for this T.

Just to make sure that I am not missing anything here, are you thinking of adding a clearer explanation to the beginning of the error message, like..
```
msg = (
     "The provided `mp` does not appear to be the matrix profile of `T`, `m`, and `k` due to the shape mismatch. The shape of `mp` must match ..."
)
```

Comment (NimaSarajpoor) [reply stumpy/stumpi.py]: @seanlaw 
> are you thinking of adding a clearer explanation to the beginning of the error message

If needed, I can revise and push again.

File: tests/test_aampi.py
Comment (NimaSarajpoor) [inline tests/test_aampi.py:1169]: No sure if this is a good way to test the new param `mp`. I pass `mp` to both naive and performant `floss`. Alternatively, we can just pass it to performant `floss`, and then compare it with naive version when `mp` is computed internally.

**Review by NimaSarajpoor — COMMENTED** [review]
@seanlaw 
Do you think I should revise the test functions? So, as we discussed before, I can remove any comparision between matrix profile indices unless they both follow `row_wise==True` approach in computing/updating matrix profile

**Comment by seanlaw** [comment]
> Do you think I should revise the test functions? So, as we discussed before, I can remove any comparision between matrix profile indices unless they both follow row_wise==True approach in computing/updating matrix profile

Whenever possible, I like to keep the comparison of matrix profile indices. For the existing tests, I don't think anything should be affected. It is only when we have an array that actually contains constant subsequences where comparisons of the indices may be a problem. Only in those cases (where there is a constant subsequence), then maybe you just pass in the pre-computed `mp`. So we test BOTH pre-computed `mp` along with `T_subseq_isconstant` at the same time for the same unit tests. It's not ideal but I am okay with it as long as we get full coverage

**Comment by NimaSarajpoor** [comment]
> Only in those cases (where there is a constant subsequence), then maybe you just pass in the pre-computed `mp`. So we test BOTH pre-computed `mp` along with `T_subseq_isconstant` at the same time for the same unit tests. It's not ideal but I am okay with it as long as we get full coverage

Correct! As long as there is only one index with global minima, we should be fine when we compare matrix profile indices.

Okay....so, for the sake of this PR only, I believe the added test functions should suffice as we get full coverage for now. Then, after merging this PR, we can resume the work on PR #817, where we add test functions for `T_subseq_isconstant`

File: stumpy/aampi.py
Comment (seanlaw) [inline stumpy/aampi.py:132]: Can you please add " but " (note the space after "but")

File: stumpy/stumpi.py
Comment (seanlaw) [inline stumpy/stumpi.py:165]: Please add " but " at the end of line 165

File: tests/test_aampi.py
Comment (seanlaw) [reply tests/test_aampi.py:1169]: Passing `mp` to both seems reasonable

**Review by seanlaw — CHANGES_REQUESTED** [review]
@NimaSarajpoor A couple of minor comments and then I think we are ready to merge when you agree.

File: stumpy/aampi.py
Comment (seanlaw) [reply stumpy/aampi.py:132]:
```
f"The shape of `mp` must match ({len(T)-m+1}, {2 * k + 2}) but "
```

Otherwise, there is no space between this line and the next line

Comment (NimaSarajpoor) [reply stumpy/aampi.py]: I forgot adding dot or space. Thanks for pointing this out! This should be fixed now.

**Pull request merged**
